### ISSUE TITLE
Force users to respect Wikipedia's User Agent policy

### ISSUE THREAD

**Issue by XioNoX** [mwclient/mwclient#302]
(I'm a SRE at the Wikimedia Foundation)

We regularly get contacted by people getting rate limited while using the API, either directly or for example using your library.
Our API also gets occasionally and accidentally abused.

To help with those 2 issues, we require API users to define a non-default User-Agent, explained in this document: https://meta.wikimedia.org/wiki/User-Agent_policy

I'm wondering if it would be possible to add documentation in this regard to your library, as well as make the UA a mandatory configuration option.

You can see an example implementation on https://github.com/martin-majlis/Wikipedia-API/issues/63

Thanks

**Comment by marcfrederick** [comment]
I've added a new warning to the documentation and updated the initial examples/quickstart to set the `clients_useragent` in #306. I'm personally undecided about whether the user agent should be a mandatory field, as this would introduce a breaking change to the library and is not something that's actually required on all wikis.

**Comment by AdamWill** [comment]
Yeah, I have the same hesitation. For instance, my own use of mwclient is for the Fedora wiki, not Wikipedia, and that does not require a user agent.

**Comment by XioNoX** [comment]
Thanks for your answers!
It would be interesting to know the feedback from Wikipedia-API about the breaking change. But as it's something easy to implement on the user side, I don't see it as a blocker (doesn't require to do significant code change to be compliant).
From experience, users often don't follow recommendations from the documentation (at best do copy/paste).
Even if a wiki doesn't require a specific user agent, it seems to me that setting one is being "a good internet citizen".
To ease the breaking change side, one option could be to show a warning (like Python's famous `DeprecationWarning`) for some time.
Whatever direction you're going, thank you for looking and thinking about this ticket, it's appreciated.

**Comment by martin-majlis** [comment]
I have made the change in June 2021. ChatGPT has knowledge cut-off in September 2021 and is not aware of this backward incompatible change. 🙈 I should have made that change immediately. :) 

I have at least made the error messages more user friendly - https://github.com/martin-majlis/Wikipedia-API/pull/256.

**Comment by andrei-korshikov** [comment]
https://meta.wikimedia.org/wiki/User-Agent_policy reads:

> This page has been moved to "[Policy:User-Agent policy](https://foundation.wikimedia.org/wiki/Policy:Wikimedia_Foundation_User-Agent_Policy)"

I think that link in
https://github.com/mwclient/mwclient/blob/24c1b418e3f2681af39d28717dbe8cbfa6e3deac/docs/source/user/connecting.rst?plain=1#L69 should be changed accordingly.

---

Also, to my understanding, `site.clients_useragent` in
https://github.com/mwclient/mwclient/blob/24c1b418e3f2681af39d28717dbe8cbfa6e3deac/docs/source/user/connecting.rst?plain=1#L66
should be changed to `site.connection.headers['User-Agent']`, see https://github.com/mwclient/mwclient/blob/24c1b418e3f2681af39d28717dbe8cbfa6e3deac/mwclient/client.py#L184 i.e. we don't have `site.clients_useragent` at all.

Sample output can be more current, too—i.e. it should contain githab link:
```
'MyCoolTool/0.2 (xyz@example.org) mwclient/0.11.0 (https://github.com/mwclient/mwclient)'
```
because of https://github.com/mwclient/mwclient/blob/24c1b418e3f2681af39d28717dbe8cbfa6e3deac/mwclient/client.py#L22

**Comment by AdamWill** [comment]
I've sent #406 to correct the issues @andrei-korshikov noted (thanks!)

I've also sent #407 which implements the original request *for wikimedia sites*, as derived from the list at https://meta.wikimedia.org/wiki/Complete_list_of_Wikimedia_projects . I'm still in two minds over whether it's a good idea to really do this, though. I figured I'd write the PR so we have something concrete to look at and decide. I've marked it as a draft for now. Thoughts welcome.

**Comment by RheingoldRiver** [comment]
After [some prolonged discussion](https://github.com/mwclient/mwclient/pull/407) and also talking a bit at the Milan hackathon, I think the conclusion is we will not add a forced UA to the client library, that puts too much maintenance burden on us. However, I am editing this ticket to make sure that if possible we correctly forward 403s caused by missing UA, which I'm not sure if we currently do.
